### ISSUE TITLE
[spirv] Suboptimal code generation for groupshared struct member functions/functions accepting groupshared arrays

### ISSUE THREAD

**Issue by pmichniewski** [microsoft/DirectXShaderCompiler#1977]
Currently DXC in SPIRV mode generates suboptimal code for member functions of groupshared structs and for functions that take groupshared arrays as params.

Testcase:
```
#define TILE_GROUP_SIZE_X 8
#define TILE_GROUP_SIZE_Y 8

struct testStruct
{
	float Data[1 + TILE_GROUP_SIZE_X + 1][1 + TILE_GROUP_SIZE_X + 1];

	float3x3 testFunc(uint topLeftX, uint topLeftY)
	{
		float3x3 toRet;

		toRet[0][0] = Data[topLeftX][topLeftY];
		toRet[0][1] = Data[topLeftX][topLeftY+1];
		toRet[0][2] = Data[topLeftX][topLeftY+2];

		toRet[1][0] = Data[topLeftX+1][topLeftY];
		toRet[1][1] = Data[topLeftX+1][topLeftY+1];
		toRet[1][2] = Data[topLeftX+1][topLeftY+2];

		toRet[2][0] = Data[topLeftX+2][topLeftY];
		toRet[2][1] = Data[topLeftX+2][topLeftY+1];
		toRet[2][2] = Data[topLeftX+2][topLeftY+2];

		return toRet;
	}
};

groupshared testStruct sharedStruct;

RWStructuredBuffer<float> testOut;

[numthreads(TILE_GROUP_SIZE_X, TILE_GROUP_SIZE_Y, 1)]
void testMain( uint3 DispatchThreadId	: SV_DispatchThreadID,
			   uint3 GroupThreadID : SV_GroupThreadID,
			   uint3 GroupID : SV_GroupID )
{
    sharedStruct.Data[GroupThreadID.x][GroupThreadID.y] = (float)GroupID.x;

    GroupMemoryBarrierWithGroupSync();

    float3x3 result = sharedStruct.testFunc( GroupThreadID.x, GroupThreadID.y );
    testOut[GroupThreadID.x] = result[0][0];
}
```

This causes a temporary variable to be created each for time testFunc is called, which in turn causes a lot of copies to be executed at runtime (50x ds_read2_b32 + 100x buffer_store_dword + 50x ds_write2_b32 are emitted by the GCN compiler).

**Comment by ehsannas** [comment]
@jaebaek Please a take a look.

**Comment by jaebaek** [comment]
@pmichniewski Thank you for opening this issue. Please help me to understand what the exact problem is.

As far as I understand, passing `groupshared` variable to a function results in creation of a new temp variable that causes unnecessary store/load because we can actually inline the function call and reuse the variable for the parameter instead of creating new one and copy the value. Am I right?

When I ran `$ dxc.exe -T cs_6_0 -E testMain -spirv test\hlsl\groupshared.hlsl -Fc test\hlsl\groupshared.hlsl.spv -fgcl`, I checked the unnecessary temp variable.

However, if I enabled the optimization i.e., running `$ dxc.exe -T cs_6_0 -E testMain -spirv test\hlsl\groupshared.hlsl -Fc test\hlsl\groupshared.hlsl.opt.spv`, it emitted the following code:

```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 43
; Schema: 0
               OpCapability Shader
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %testMain "testMain" %gl_GlobalInvocationID %gl_LocalInvocationID %gl_WorkGroupID
               OpExecutionMode %testMain LocalSize 8 8 1
               OpSource HLSL 600
               OpName %testStruct "testStruct"
               OpMemberName %testStruct 0 "Data"
               OpName %sharedStruct "sharedStruct"
               OpName %type_RWStructuredBuffer_float "type.RWStructuredBuffer.float"
               OpName %testOut "testOut"
               OpName %testMain "testMain"
               OpDecorate %gl_GlobalInvocationID BuiltIn GlobalInvocationId
               OpDecorate %gl_LocalInvocationID BuiltIn LocalInvocationId
               OpDecorate %gl_WorkGroupID BuiltIn WorkgroupId
               OpDecorate %testOut DescriptorSet 0
               OpDecorate %testOut Binding 0
               OpDecorate %_runtimearr_float ArrayStride 4
               OpMemberDecorate %type_RWStructuredBuffer_float 0 Offset 0
               OpDecorate %type_RWStructuredBuffer_float BufferBlock
        %int = OpTypeInt 32 1
      %int_0 = OpConstant %int 0
       %uint = OpTypeInt 32 0
     %uint_2 = OpConstant %uint 2
    %uint_10 = OpConstant %uint 10
      %float = OpTypeFloat 32
%_arr_float_uint_10 = OpTypeArray %float %uint_10
%_arr__arr_float_uint_10_uint_10 = OpTypeArray %_arr_float_uint_10 %uint_10
 %testStruct = OpTypeStruct %_arr__arr_float_uint_10_uint_10
%_ptr_Workgroup_testStruct = OpTypePointer Workgroup %testStruct
%_runtimearr_float = OpTypeRuntimeArray %float
%type_RWStructuredBuffer_float = OpTypeStruct %_runtimearr_float
%_ptr_Uniform_type_RWStructuredBuffer_float = OpTypePointer Uniform %type_RWStructuredBuffer_float
     %v3uint = OpTypeVector %uint 3
%_ptr_Input_v3uint = OpTypePointer Input %v3uint
       %void = OpTypeVoid
         %23 = OpTypeFunction %void
%_ptr_Workgroup_float = OpTypePointer Workgroup %float
   %uint_264 = OpConstant %uint 264
%_ptr_Function_float = OpTypePointer Function %float
%_ptr_Uniform_float = OpTypePointer Uniform %float
%sharedStruct = OpVariable %_ptr_Workgroup_testStruct Workgroup
    %testOut = OpVariable %_ptr_Uniform_type_RWStructuredBuffer_float Uniform
%gl_GlobalInvocationID = OpVariable %_ptr_Input_v3uint Input
%gl_LocalInvocationID = OpVariable %_ptr_Input_v3uint Input
%gl_WorkGroupID = OpVariable %_ptr_Input_v3uint Input
%_ptr_Function__arr__arr_float_uint_10_uint_10 = OpTypePointer Function %_arr__arr_float_uint_10_uint_10
   %testMain = OpFunction %void None %23
         %29 = OpLabel
         %30 = OpVariable %_ptr_Function__arr__arr_float_uint_10_uint_10 Function
         %31 = OpLoad %v3uint %gl_LocalInvocationID
         %32 = OpLoad %v3uint %gl_WorkGroupID
         %33 = OpCompositeExtract %uint %32 0
         %34 = OpConvertUToF %float %33
         %35 = OpCompositeExtract %uint %31 0
         %36 = OpCompositeExtract %uint %31 1
         %37 = OpAccessChain %_ptr_Workgroup_float %sharedStruct %int_0 %35 %36
               OpStore %37 %34
               OpControlBarrier %uint_2 %uint_2 %uint_264
         %38 = OpLoad %testStruct %sharedStruct
         %39 = OpCompositeExtract %_arr__arr_float_uint_10_uint_10 %38 0
               OpStore %30 %39
         %40 = OpAccessChain %_ptr_Function_float %30 %35 %36
         %41 = OpLoad %float %40
               OpStore %sharedStruct %38
         %42 = OpAccessChain %_ptr_Uniform_float %testOut %int_0 %35
               OpStore %42 %41
               OpReturn
               OpFunctionEnd
```

which inlined the method i.e., `testFunc()` and removed unneccessary temp variable.

As far as I understand, this is done by

1. Inlining based on `InlineExhaustivePass` of external/SPIRV-Tools/source/opt/inline_exhaustive_pass.cpp
2. Local store/load elimination by `LocalSingleBlockLoadStoreElimPass` of external/SPIRV-Tools/source/opt/local_single_block_elim_pass.cpp

I guess if you run spirv-opt, it would be fine.
Please let me know if I misunderstood something.

**Comment by ehsannas** [comment]
Just to clarify, you don't need to run spirv-opt manually. DXC's SPIR-V code generator runs it by default under the hood.

**Comment by pmichniewski** [comment]
The issue is caused by these instructions:
```
         %38 = OpLoad %testStruct %sharedStruct
         %39 = OpCompositeExtract %_arr__arr_float_uint_10_uint_10 %38 0
               OpStore %30 %39
```

I copied the SPIRV assembly you included, assembled it with spirv-as and decompiled using spirv-cross and this is what I got:

```
#version 450
layout(local_size_x = 8, local_size_y = 8, local_size_z = 1) in;

struct testStruct
{
    float Data[10][10];
};

layout(binding = 0, std430) buffer type_RWStructuredBuffer_float
{
    float _m0[];
} testOut;

shared testStruct sharedStruct;

void main()
{
    sharedStruct.Data[gl_LocalInvocationID.x][gl_LocalInvocationID.y] = float(gl_WorkGroupID.x);
    memoryBarrierShared();
    barrier();
    float _30[10][10] = sharedStruct.Data; // <--- this causes the unnecessary copy
    sharedStruct = sharedStruct;
    testOut._m0[gl_LocalInvocationID.x] = _30[gl_LocalInvocationID.x][gl_LocalInvocationID.y];
}
```
Passing this through AMD's rga results in the following:
```
    9 |   5 | ::v^^               | ds_read2_b32 v[3:4], v2
   10 |   7 | ::v::^^             | ds_read2_b32 v[5:6], v2
   11 |   9 | ::v::::^^           | ds_read2_b32 v[7:8], v2
   12 |  11 | ::v::::::^^         | ds_read2_b32 v[9:10], v2
   13 |  13 | ::v::::::::^^       | ds_read2_b32 v[11:12], v2
   14 |  15 | ::v::::::::::^^     | ds_read2_b32 v[13:14], v2
   15 |  17 | ::v::::::::::::^^   | ds_read2_b32 v[15:16], v2
   16 |  17 | :::::::::::::::::   | s_waitcnt lgkmcnt(6)
   17 |  18 | :::v::::::::::::: ^ | v_mov_b32 v18, v3
   18 |  17 | ::: v:::::::::::: : | v_mov_b32 v19, v4
   19 |  16 | :::  :::::::::::: : | s_waitcnt lgkmcnt(5)
   20 |  16 | :::  v::::::::::: : | v_mov_b32 v20, v5
   21 |  15 | :::   v:::::::::: : | v_mov_b32 v21, v6
   22 |  14 | :::    :::::::::: : | s_waitcnt lgkmcnt(4)
   23 |  14 | :::    v::::::::: : | v_mov_b32 v22, v7
   24 |  13 | :::     v:::::::: : | v_mov_b32 v23, v8
   25 |  12 | :::      :::::::: : | s_waitcnt lgkmcnt(3)
   26 |  12 | :::      v::::::: : | v_mov_b32 v24, v9
   27 |  11 | :::       v:::::: : | v_mov_b32 v25, v10
   28 |  10 | :::        :::::: : | s_waitcnt lgkmcnt(2)
   29 |  10 | :::        v::::: : | v_mov_b32 v26, v11
   30 |   9 | :::         v:::: : | v_mov_b32 v27, v12
   31 |   8 | :::          :::: : | s_waitcnt lgkmcnt(1)
   32 |   8 | :::          v::: : | v_mov_b32 v28, v13
   33 |   7 | :::           v:: : | v_mov_b32 v29, v14
   34 |   6 | :::            :: : | s_waitcnt lgkmcnt(0)
   35 |   6 | :::            v: : | v_mov_b32 v30, v15
   36 |   5 | :::             v : | v_mov_b32 v31, v16
```
which is repeated 7x for a total of ~100 copy operations. My original shader had even worse code generated (it had 5 instances of the struct resulting in 5 x 200 copy instructions, using 100+ registers for this) but I cannot share it here.

**Comment by ehsannas** [comment]
Thanks for the additional info.

@jaebaek We should investigate a bit more closely. The mismatch between the `Function` storage class and `Workgroup` storage class could result in a load from the source storage class followed by a store in the destination storage class.

**Comment by jaebaek** [comment]
Sure, let me take a look. @pmichniewski Thank you for explaining it in detail.

**Comment by jaebaek** [comment]
Fixed by #1991
